src/Compilers/Core/Portable/SourceGeneration/IncrementalGeneratorSyntaxWalker.cs

```diff
@@ -0,0 +1,61 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Text;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis
+{
+    internal sealed class IncrementalGeneratorSyntaxWalker : SyntaxWalker
+    {
+        private readonly ArrayBuilder<ISyntaxTransformBuilder> _builders;
+
+        private readonly ArrayBuilder<ArrayBuilder<SyntaxNode>?> _results;
+
+        internal IncrementalGeneratorSyntaxWalker(ArrayBuilder<ISyntaxTransformBuilder> builders, ArrayBuilder<ArrayBuilder<SyntaxNode>?> results)
+        {
+            _builders = builders;
+            _results = results;
+        }
+
+        public static void VisitNodeForBuilders(SyntaxNode root, SemanticModel semanticModel, ArrayBuilder<ISyntaxTransformBuilder> builders)
```

Comment:
Consider making this a local function inside `GeneratorDriver.RunGeneratorsCore`

Reply:
Chuck suggested I inline more of it here :p I think this is probably going to get edited as part of the input cleanup anyway, so I'll leave as-is for now.

---
In reply to: [627075356](https://github.com/dotnet/roslyn/pull/53147#discussion_r627075356) [](ancestors = 627075356)